pkg/controllers/leaderworkerset_controller.go

```diff
@@ -206,77 +206,111 @@ func SetupIndexes(indexer client.FieldIndexer) error {
 //     together with scaling down.
 //
 // At rest, Partition should always be zero.
-func (r *LeaderWorkerSetReconciler) rollingUpdatePartition(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (int32, error) {
+//
+// For Replicas:
+//   - When rolling update, Replicas is equal to (spec.Replicas+maxSurge)
+//   - Otherwise, Replicas is equal to spec.Replicas
+//   - One exception here is when (Partition+1) equals to MaxSurge, we should reclaim the extra replicas gradually.
+//
+// If leaderWorkerSet is under rolling update, a new rolling update will rebuild the bursted replicas
+// and not waiting for the former group ready.
+func (r *LeaderWorkerSetReconciler) rollingUpdateParameters(ctx context.Context, lws *leaderworkerset.LeaderWorkerSet) (partition int32, replicas int32, err error) {
+	log := ctrl.LoggerFrom(ctx)
+	lwsReplicas := *lws.Spec.Replicas
+
 	sts := &appsv1.StatefulSet{}
-	err := r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
+	err = r.Get(ctx, types.NamespacedName{Name: lws.Name, Namespace: lws.Namespace}, sts)
 	if err != nil {
-		// If sts not created yet, all partitions should be updated.
+		// If sts not created yet, all partitions should be updated,
+		// replicas should not change.
 		if apierrors.IsNotFound(err) {
-			return 0, nil
+			return 0, lwsReplicas, nil
 		}
-		return 0, err
+		return 0, 0, err
+	}
+
+	partition = *sts.Spec.UpdateStrategy.RollingUpdate.Partition
+	rollingStep, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(lwsReplicas), false)
+	if err != nil {
+		return 0, 0, err
+	}
+	maxSurge, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxSurge, int(lwsReplicas), false)
+	if err != nil {
+		return 0, 0, err
 	}
 
-	partition := *sts.Spec.UpdateStrategy.RollingUpdate.Partition
-	replicas := *lws.Spec.Replicas
-	rollingStep, err := intstr.GetValueFromIntOrPercent(&lws.Spec.RolloutStrategy.RollingUpdateConfiguration.MaxUnavailable, int(replicas), false)
+	// Scale down the bursted replicas gradually.
+	defer func() {
+		replicas = tunedStsReplicas(replicas, partition, int32(maxSurge))
+		log.Info("rollingUpdate parameters", "replicas", replicas, "partition", partition)
+	}()
+
+	burstReplicas := lwsReplicas + int32(maxSurge)
+	stsReplicas, err := strconv.Atoi(sts.Annotations[leaderworkerset.ReplicasAnnotationKey])
 	if err != nil {
-		return 0, err
+		return 0, 0, err
 	}
 
+	duringRollingUpdate := partition != 0
+	replicasUpdated := stsReplicas != int(*lws.Spec.Replicas)
+
 	// Indicates a new rolling update here.
 	if templateUpdated(sts, lws) {
-		// Processing scaling up/down first prior to rolling update.
-		if replicasUpdated(sts, lws) {
-			return min(*lws.Spec.Replicas, *sts.Spec.Replicas), nil
+		if replicasUpdated {
+			// Processing scaling up/down first prior to rolling update.
+			return min(lwsReplicas, int32(stsReplicas)), burstReplicas, nil
```

Comment:
When MaxSurge is specified, and *lws.spec.replica changed, we scale lws to *lws.spec.replicas + maxSurge right? The partition here is the number of replicas needs to be updated considering scale up/down. And each step we upgrade MaxUnavailable number of replicas.

Reply:
this is actually slightly different from what I imaged. We have the SurgeUpgrade on GKE which we upgrade MaxSurge+MaxUnavailable number of nodes each time. so rollingStep = MaxSurge+MaxUnavailable. @ahg-g do you think we should do sth similar here as well? Otherwise it feels to me this feature more looks like enabling burst replicas during upgrade

Reply:
Make sense to me, also aligned with the comment: `followed by scaling up the new replicas, ensuring
	// that at least 70% of original number of replicas are available at all times
	// during the update.`